### PR TITLE
[CEF] Add new port (#332)

### PR DESCRIPTION
**Pull request by Pospelove**
Only `x86-windows` is currently supported.

See [cef_vcpkg_demo.zip](https://github.com/microsoft/vcpkg/files/4212259/cef_vcpkg_demo.zip) to test this PR with CEF sample apps (use `gen.bat`).

### CODE REVIEW

File: ports/cef/portfile.cmake [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+else()
+    message(FATAL_ERROR "This port is not supported on the current platform")
+endif()
+
+if (VCPKG_TARGET_ARCHITECTURE STREQUAL "x64")
+    message(FATAL_ERROR "This port is not supported on the current platform")
+else()
+    set(architecture 32)
+endif()
+
```
Comment (NancyLi1013) [inline ports/cef/portfile.cmake]: These codes might be updated as `vcpkg_fail_port_install(ON_ARCH "x64" "arm" ON_TARGET "Linux" "osx" "uwp" )`.

File: ports/cef/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    else()
+        set(sys SysWoW64)
+    endif()
+
+    if (NOT EXISTS $ENV{SystemRoot}/${sys}/d3dcompiler_47.dll)
+        message(FATAL_ERROR "d3dcompiler_47.dll is missing")
+    endif()
+endif()
+
+file(INSTALL ${CEF_PATH}/LICENSE.txt DESTINATION ${CURRENT_PACKAGES_DIR}/share/cef RENAME copyright)
```
Comment (NancyLi1013) [inline ports/cef/portfile.cmake]: Could you please update 
`file(INSTALL ${CEF_PATH}/LICENSE.txt DESTINATION ${CURRENT_PACKAGES_DIR}/share/cef RENAME copyright)`
as
`file(INSTALL ${CEF_PATH}/LICENSE.txt DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT} RENAME copyright)`?

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

**Comment by Pospelove** [comment]
@NancyLi1013, @JackBoosY I see "waiting for response" label on this PR which means "Waiting for the PR author to respond to this PR". So what should I do? Please let me know if I didn't do the requested changes correctly.

**Comment by JackBoosY** [comment]
It seems have some regressions, please get the log here([windows](https://dev.azure.com/vcpkg/public/_build/results?buildId=29388&view=artifacts&type=publishedArtifacts), [linux](https://dev.azure.com/vcpkg/public/_build/results?buildId=29389&view=artifacts&type=publishedArtifacts), [osx](https://dev.azure.com/vcpkg/public/_build/results?buildId=29390&view=artifacts&type=publishedArtifacts)).

**Comment by Pospelove** [comment]
How to exclude unsupported architectures/platforms from CI?
There are 
```
CMake Error at scripts/cmake/vcpkg_fail_port_install.cmake:93 (message):
  Architecture 'arm' not supported by cef!

Call Stack (most recent call first):
  ports/cef/portfile.cmake:3 (vcpkg_fail_port_install)
  scripts/ports.cmake:90 (include)
```

How do I see what actually goes wrong here? I didn't see this in artifacts. I also tried to run dumpbin.exe on my local machine and it finished successfully.
```
2020-02-20T02:03:39.7934388Z -- Performing post-build validation
2020-02-20T02:03:42.7328803Z Running command:
2020-02-20T02:03:42.7330391Z    "C:\VS2017\Stable\VC\Tools\MSVC\14.22.27905\bin\HostX86\x86\dumpbin.exe" /exports "C:\vsts\_work\3\s\packages\cef_x86-windows\debug\bin\cef\libcef.dll"
2020-02-20T02:03:42.7331749Z  failed
2020-02-20T02:03:43.0497498Z CI test is complete
2020-02-20T02:03:43.0504464Z CI test script is complete
2020-02-20T02:03:43.1854579Z ##[error]PowerShell exited with code '1'.
2020-02-20T02:03:43.2271273Z ##[section]Finishing: ** Build vcpkg and test ports **
2020-02-20T02:03:43.2315814Z ##[section]Starting: Publish Artifact: x86-windows port build failure logs
```

**Comment by JackBoosY** [comment]
If cef doesn't support some triplets. you should add `cef:<triplet>=fail` to _VCPKG_PATH/scripts/ci.baseline.txt_.

**Comment by Pospelove** [comment]
Edited ci.baseline.txt. So what about dumpbin.exe failure?

**Comment by JackBoosY** [comment]
I found some questions:
1. Generate both dynamic and static libraries when building dynamically。
2. Some static libraries should be moved from _bin_ to _lib_.
3. Path _bin_ should not store non-dynamic library files.
4. `egl` libraries should not be stored.

**Comment by Pospelove** [comment]
> I found some questions:
> 
> 1. Generate both dynamic and static libraries when building dynamically。
> 2. Some static libraries should be moved from _bin_ to _lib_.
> 3. Path _bin_ should not store non-dynamic library files.
> 4. `egl` libraries should not be stored.

1. CEF is basically distributed as `libcef.lib` and it's binary dependencies. So I am not sure should it be treated as static or dynamic. Theoretically, we could build CEF and all its dependencies from sources into a single .dll or .lib. But I think building the Chromium on the user's machine is not an option. I also do not sure Chromium supports distributing all its dependencies as static libraries.
2. It's reasonable. The reason I placed them into the `bin` folder was my thought that CEF users are relying on the folder structure. But I reviewed `cef_variables.cmake` again and I now don't think that can break something. However, I receive an error on the post-build validation step: `Could not find proper second linker member` after changes fixing this.
3. OK, I'll try to fix that.
4. So, how should we work with EGL libraries? Adding a new port?

**Comment by JackBoosY** [comment]
> CEF is basically distributed as libcef.lib and it's binary dependencies. So I am not sure should it be treated as static or dynamic. Theoretically, we could build CEF and all its dependencies from sources into a single .dll or .lib. But I think building the Chromium on the user's machine is not an option. I also do not sure Chromium supports distributing all its dependencies as static libraries.

Just confirm that these libraries will be built as static or dynamic libraries.

> It's reasonable. The reason I placed them into the bin folder was my thought that CEF users are relying on the folder structure. But I reviewed cef_variables.cmake again and I now don't think that can break something. However, I receive an error on the post-build validation step: Could not find proper second linker member after changes fixing this.

We must put the static libraries in the _lib_ root directory and the dynamic libraries in the _bin_ root directory to ensure that vcpkg finally checks for these libraries. Please correct the paths in cmake based on their directories.

> So, how should we work with EGL libraries? Adding a new port?

The port `qt5-base` also generates EGL libraries, so I think the best way to handle this is to add a new port.

**Comment by Pospelove** [comment]
> Just confirm that these libraries will be built as static or dynamic libraries.

There are prebuilt both dynamic and static libraries in the CEF distribution. How do I handle them?

**Comment by JackBoosY** [comment]
@Pospelove I'm afraid we can't use prebuild, vcpkg requires all binaries to be generated by code. Can you modify it?

**Comment by Pospelove** [comment]
@JackBoosY  Got it. What about [this](https://docs.microsoft.com/en-us/windows/win32/directx-sdk--august-2009-)?

> Any project that uses run-time shader compilation must have D3DCOMPILER_xx.DLL copied to the local executable path for the project. This DLL is available in this sub-directory of the Windows SDK installation under %ProgramFiles(x86)%\Windows Kits\8.0\Redist\D3D\ or %ProgramFiles(x86)%\Windows Kits\8.1\Redist\D3D\ where is x86 and x64.
> 
> The D3DCOMPILER_46.DLL or D3DCOMPILER_47.DLL from the Windows SDK is not a system component and should not be copied to the Windows system directory. You can redistribute this DLL to other computers with your application as a side-by-side DLL.

**Comment by JackBoosY** [comment]
@Pospelove I don't understand what you mean.
Only SDKs can use prebuild, as they are installed manually by the user.

**Comment by Pospelove** [comment]
Well. So, I'll take a look at this later to modify this to build all on the user's machine. I found out that DirectX SDK can be handled like in #8677. I also think this PR requires chromium and egl ports to be added.

**Comment by Voskrese** [comment]
> The D3DCOMPILER_46.DLL or D3DCOMPILER_47.DLL from the Windows SDK is not a system component and should not be copied to the Windows system directory. You can redistribute this DLL to other computers with your application as a side-by-side DLL.

**C:\Program Files (x86)\Windows Kits\10\Redist\D3D\x64** >`tree /F`
```
    d3dcompiler_47.dll
    d3dcsx_47.dll
    dxcompiler.dll
    dxil.dll
```

**Comment by JackBoosY** [comment]
Any progress?

**Comment by JackBoosY** [comment]
Closing this PR since it seems that no progress is being made. Please reopen if work is still being done.

**Pull request closed without merging**